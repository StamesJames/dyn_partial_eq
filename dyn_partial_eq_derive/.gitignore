Cargo.lock

target/
.vscode/

**/*.rs.bk

*.pdb