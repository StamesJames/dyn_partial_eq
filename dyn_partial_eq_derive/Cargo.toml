[package]
name = "dyn_partial_eq_derive"
version = "0.1.0"
edition = "2021"
license = "MIT"
keywords = ["dynamic", "derive"]
description = "This crate enables you to derive PertialEq for types that hold dyn objects by using the  DynPartialEq from the associated crate dyn_partial_eq"
readme = "README.md"

[lib]
proc-macro = true

[dependencies]
proc-macro2 = "1.0.60"
quote = "1.0.28"
syn = "2.0.18"
