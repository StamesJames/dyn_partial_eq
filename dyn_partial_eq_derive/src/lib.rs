use proc_macro::TokenStream;
mod as_any_derive;
mod dyn_partial_eq_derive;
mod partial_eq_dyn_derive;

/// This derives the AsAny Trait in its most easy form of an identity function
#[proc_macro_derive(AsAny)]
pub fn as_any_derive(input: TokenStream) -> TokenStream {
    let ast = as_any_derive::parse(input.into());
    as_any_derive::impl_as_any(&ast).into()
}

/// This derives the DynPartialEq Trait by simple downcasting the Any Object to the given type and calling the regular PartialEq comparisson. Therefor your Type has to implement PartialEq
#[proc_macro_derive(DynPartialEq)]
pub fn dyn_partial_eq_derive(input: TokenStream) -> TokenStream {
    let ast = dyn_partial_eq_derive::parse(input.into());
    dyn_partial_eq_derive::impl_dyn_partial_eq(&ast).into()
}

/// This derives PartialEq but manages Trait Objects by calling their dyn_eq methods and casting. To work all Trait Objects have to be boundet by DynPartialEq and AsAny.
#[proc_macro_derive(PartialEqDyn)]
pub fn partial_eq_any_eq_derive(input: TokenStream) -> TokenStream {
    let ast = partial_eq_dyn_derive::parse(input.into());
    partial_eq_dyn_derive::impl_partial_eq_dyn(&ast).into()
}

#[cfg(test)]
mod test {
    use crate::{as_any_derive, dyn_partial_eq_derive, partial_eq_dyn_derive};
    use quote::quote;
    #[test]
    fn parse_named_struct() {
        as_any_derive::parse(quote! {
            struct NamedStruct {
                field1: i32,
                field2: String,
                field3: Box<i32>,
            }
        });
        dyn_partial_eq_derive::parse(quote! {
            struct NamedStruct {
                field1: i32,
                field2: String,
                field3: Box<i32>,
            }
        });
        partial_eq_dyn_derive::parse(quote! {
            struct NamedStruct {
                field1: i32,
                field2: String,
                field3: Box<i32>,
            }
        });
    }
    #[test]
    fn parse_unnamed_struct() {
        as_any_derive::parse(quote! {
            struct UnnamedStruct (
                i32,
                String,
                Box<i32>,
            );
        });
        dyn_partial_eq_derive::parse(quote! {
            struct UnnamedStruct (
                i32,
                String,
                Box<i32>,
            );
        });
        partial_eq_dyn_derive::parse(quote! {
            struct UnnamedStruct (
                i32,
                String,
                Box<i32>,
            );
        });
    }
    #[test]
    fn parse_named_enum() {
        as_any_derive::parse(quote! {
            enum NamedEnum {
                First{first:i32},
                Second{sec_first:String,
                sec_second:Box<i32>,},
            }
        });
        dyn_partial_eq_derive::parse(quote! {
            enum NamedEnum {
                First{first:i32},
                Second{sec_first:String,
                sec_second:Box<i32>,},
            }
        });
        partial_eq_dyn_derive::parse(quote! {
            enum NamedEnum {
                First{first:i32},
                Second{sec_first:String,
                sec_second:Box<i32>,},
            }
        });
    }
    #[test]
    fn parse_unnamed_enum() {
        as_any_derive::parse(quote! {
            enum UnnamedEnum {
                First(i32),
                Second(String,
                Box<i32>,),
            }
        });
        dyn_partial_eq_derive::parse(quote! {
            enum UnnamedEnum {
                First(i32),
                Second(String,
                Box<i32>,),
            }
        });
        partial_eq_dyn_derive::parse(quote! {
            enum UnnamedEnum {
                First(i32),
                Second(String,
                Box<i32>,),
            }
        });
    }
    #[test]
    fn parse_mixed_named_enum() {
        as_any_derive::parse(quote! {
            enum MixedNamedEnum {
                First(i32,String),
                Second{first:String,
                second:Box<i32>,},
            }
        });
        dyn_partial_eq_derive::parse(quote! {
            enum MixedNamedEnum {
                First(i32,String),
                Second{first:String,
                second:Box<i32>,},
            }
        });
        partial_eq_dyn_derive::parse(quote! {
            enum MixedNamedEnum {
                First(i32,String),
                Second{first:String,
                second:Box<i32>,},
            }
        });
    }
}
