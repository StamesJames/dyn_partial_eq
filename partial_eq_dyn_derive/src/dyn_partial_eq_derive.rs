use proc_macro2::TokenStream;
use quote::quote;
use syn::DeriveInput;

pub fn parse(input: TokenStream) -> DeriveInput {
    syn::parse2(input).expect("The Item is supposed to be a Struct or Enum")
}

pub fn impl_dyn_partial_eq(ast: &syn::DeriveInput) -> TokenStream {
    let item_name = &ast.ident;
    quote! {
        impl partial_eq_dyn::DynPartialEq for #item_name {
            fn dyn_eq(&self, other: &dyn std::any::Any) -> bool {
                other
                    .downcast_ref::<#item_name>()
                    .map_or(false, |other| self == other)
            }
        }
    }
}
