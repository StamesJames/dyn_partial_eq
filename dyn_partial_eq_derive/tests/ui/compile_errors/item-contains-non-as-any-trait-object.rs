use dyn_partial_eq::DynPartialEq;
use dyn_partial_eq_derive::PartialEqDyn;

trait NonAsAny: DynPartialEq {}

#[derive(PartialEqDyn)]
struct TestStruct(Box<dyn NonAsAny>);

fn main() {}
