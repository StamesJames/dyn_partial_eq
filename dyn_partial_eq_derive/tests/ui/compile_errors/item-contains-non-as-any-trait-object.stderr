error[E0599]: no method named `as_any` found for reference `&Box<(dyn NonAsAny + 'static)>` in the current scope
 --> tests/ui/compile_errors/item-contains-non-as-any-trait-object.rs:6:10
  |
6 | #[derive(PartialEqDyn)]
  |          ^^^^^^^^^^^^ method not found in `&Box<dyn NonAsAny>`
  |
  = help: items from traits can only be used if the trait is implemented and in scope
  = note: the following trait defines an item `as_any`, perhaps you need to implement it:
          candidate #1: `AsAny`
  = note: this error originates in the derive macro `PartialEqDyn` (in Nightly builds, run with -Z macro-backtrace for more info)
